File: _Research_testing/hooktheoryDataCompatTest.mjs
import assert from "node:assert/strict";
import {
  LEGACY_PIANO_MAN_CHORUS,
  migrateLegacyChord,
  migrateLegacySectionData,
} from "../web-player/lib/hooktheoryDataCompat.js";

const oldChord = {
  root: 5,
  beat: 40,
  duration: 3,
  type: 11,
  inversion: 0,
  applied: 0,
  adds: [],
  omits: [],
  alterations: [],
  suspensions: [],
};

const oldSection = {
  numericId: LEGACY_PIANO_MAN_CHORUS.numericId,
  metadata: { fp: LEGACY_PIANO_MAN_CHORUS.fingerprint },
  chords: [oldChord],
};

const migrated = migrateLegacySectionData(oldSection);
assert.deepEqual(migrated.chords[0], { ...oldChord, type: 9, suspensions: [4] });
assert.deepEqual(oldSection.chords[0], oldChord, "migration must not mutate cached data");

const canonical = { ...oldChord, type: 9, suspensions: [4] };
assert.equal(migrateLegacyChord(canonical, oldSection), canonical, "canonical payloads stay unchanged");

const genuineEleventh = { ...oldChord, beat: 41 };
assert.equal(migrateLegacyChord(genuineEleventh, oldSection), genuineEleventh, "only the stale signature migrates");

const genuineOtherSong = { ...oldChord };
assert.equal(
  migrateLegacyChord(genuineOtherSong, { numericId: "other", metadata: { fp: "other" } }),
  genuineOtherSong,
  "genuine type=11 chords from other sources stay unchanged",
);

console.log("Hooktheory data compatibility tests passed.");

File: android/app/src/test/java/com/inquiring/android/HooktheoryDataCompatTest.kt
package com.inquiring.android

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.intOrNull
import org.junit.Assert.assertEquals
import org.junit.Assert.assertSame
import org.junit.Test

class HooktheoryDataCompatTest {
    private val json = Json { ignoreUnknownKeys = true }

    @Test
    fun migratesOnlyTheKnownLegacyPianoManChord() {
        val oldChord = json.decodeFromString<JsonObject>(
            """{"root":5,"beat":40,"duration":3,"type":11,"inversion":0,"applied":0,"adds":[],"omits":[],"alterations":[],"suspensions":[]}"""
        )
        val section = ExtractedSection(
            numericId = kotlinx.serialization.json.JsonPrimitive("1714973"),
            chords = listOf(oldChord),
            metadata = json.decodeFromString("""{"fp":"94c3b7dc6a7f8804312aae2fa40079291ec84eb95"}"""),
        )

        val migrated = HooktheoryDataCompat.migrateSection(section)
        assertEquals(9, (migrated.chords.single()["type"] as JsonPrimitive).intOrNull)
        assertEquals(
            listOf(4),
            (migrated.chords.single()["suspensions"] as JsonArray)
                .map { (it as JsonPrimitive).intOrNull },
        )
        assertEquals(11, (section.chords.single()["type"] as JsonPrimitive).intOrNull)
    }

    @Test
    fun leavesGenuineEleventhsAlone() {
        val chord = json.decodeFromString<JsonObject>(
            """{"root":1,"beat":56.5,"duration":1,"type":11,"suspensions":[]}"""
        )
        val section = ExtractedSection(
            numericId = kotlinx.serialization.json.JsonPrimitive("other"),
            chords = listOf(chord),
            metadata = json.decodeFromString("""{"fp":"other"}"""),
        )
        val migrated = HooktheoryDataCompat.migrateSection(section)
        assertSame(section, migrated)
    }
}
